File: experiments/006-SmalltalkVsJava/cells/java-gradle_1-Empty_free/2022-1c-recuperatorio-parcial1/20260915-110924-39025/output/project/src/main/java/customerimporter/ZipCode.java
package customerimporter;

import java.util.List;
import java.util.function.Function;
import java.util.function.Predicate;
import java.util.function.Supplier;

public abstract class ZipCode {

    private record ZipCodeCreator(Predicate<String> appliesTo, Function<String, ZipCode> create) {}

    private static final List<ZipCodeCreator> creators = List.of(
        new ZipCodeCreator(OldZipCode::appliesTo, OldZipCode::new),
        new ZipCodeCreator(NewZipCode::appliesTo, NewZipCode::new));

    private final String code;

    // instance creation

    public static ZipCode from(String aZipCode) {
        ZipCode newZipCode = creatorFor(aZipCode).apply(aZipCode);
        newZipCode.assertIsValid();
        return newZipCode;
    }

    // error messages

    // The tests expect this message for a zip code that is neither old nor new, so it is kept as it was
    public static String invalidZipCodeTypeErrorDescription() {
        return "Invalid identification type";
    }

    // instance creation - private

    private static Function<String, ZipCode> creatorFor(String aZipCode) {
        return creators.stream()
            .filter(aCreator -> aCreator.appliesTo().test(aZipCode))
            .findFirst()
            .orElseThrow(() -> new RuntimeException(invalidZipCodeTypeErrorDescription()))
            .create();
    }

    // initialization

    protected ZipCode(String aCode) {
        code = aCode;
    }

    // zip code

    public String code() {
        return code;
    }

    public abstract Object value();

    // type

    public boolean isOld() {
        return false;
    }

    public boolean isNew() {
        return false;
    }

    public Object oldZipCodeIfNone(Supplier<Object> aNoneBlock) {
        return aNoneBlock.get();
    }

    public Object newZipCodeIfNone(Supplier<Object> aNoneBlock) {
        return aNoneBlock.get();
    }

    // validation

    protected abstract void assertIsValid();

    protected abstract String invalidZipCodeErrorDescription();

    protected void assertThat(boolean aCondition) {
        if (!aCondition) throw new RuntimeException(invalidZipCodeErrorDescription());
    }
}

File: experiments/020-PythonVsJava/cells/python-pytest_1-Empty_free/2022-1c-recuperatorio-parcial1/20260921-205812-95418/analysis/acceptance-project/src/customerimporter/zip_code.py
from abc import ABC, abstractmethod


class ZipCode(ABC):

    _code = None

    # instance creation

    @classmethod
    def for_(cls, a_zip_code):
        for a_zip_code_class in cls.__subclasses__():
            if a_zip_code_class.is_for(a_zip_code): return a_zip_code_class.with_code(a_zip_code)
        raise RuntimeError(cls.invalid_zip_code_type_error_description())

    @classmethod
    def with_code(cls, a_zip_code):
        cls.assert_is_valid_code(a_zip_code)
        return cls(a_zip_code)

    # error messages

    @classmethod
    def invalid_zip_code_type_error_description(cls):
        # the text the tests expect for a zip code that is neither old nor new
        return "Invalid identification type"

    @classmethod
    @abstractmethod
    def invalid_code_error_description(cls):
        pass

    # initialization

    def __init__(self, a_zip_code):
        self._code = a_zip_code

    # type

    @classmethod
    @abstractmethod
    def is_for(cls, a_zip_code):
        pass

    # code - validation

    @classmethod
    @abstractmethod
    def assert_is_valid_code(cls, a_zip_code):
        pass

    @classmethod
    def assert_valid_code(cls, a_condition):
        if not a_condition: raise RuntimeError(cls.invalid_code_error_description())

    # code

    def code(self):
        return self._code

    @abstractmethod
    def new_zip_code_if_none(self, a_none_block):
        pass

    @abstractmethod
    def old_zip_code_if_none(self, a_none_block):
        pass

    @abstractmethod
    def is_new(self):
        pass

    @abstractmethod
    def is_old(self):
        pass
